### PR TITLE
Avoid unnecessary heap allocation when parsing frame headers.

### LINKED ISSUE #105: Remove unnecessary heap allocation in frame parsing.
While I was debugging #104, I spotted that we have an unnecessary heap allocation in `ByteBuffer.readFrameHeader`:

https://github.com/apple/swift-nio-http2/blob/429862635ade915661dc6cad0a970ed0993b7c56/Sources/NIOHTTP2/HTTP2FrameParser.swift#L1166

This whole method should probably be rewritten with `readWithUnsafeReadableBytes` or by using some `readInteger` calls: we can benchmark the difference between those two implementations.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

As noted in #105, I spotted that we were heap-allocating an array
when parsing the frame header. That heap allocation isn't necessary, we
can just write directly into an Int, which should save us some heap spray.

Modifications:

- Rewrote code to avoid heap allocating array and instead to use Int.
- Elided some bounds checks by using an unsafe pointer, since I was
    already here and it saved about 1% performance in a hot loop.

Result:

Fewer heap allocations on the hot code path.
Resolves #105

### CODE REVIEW

**Comment by Lukasa** [comment]
This has a fraction of one percent of a performance improvement in the real-world benchmark I'm running: just not really noticeable. Still worth avoiding the allocation, but just worth keeping it in perspective.

File: Sources/NIOHTTP2/HTTP2FrameParser.swift [hunk 1]
```diff
@@ -1163,13 +1163,14 @@ fileprivate extension ByteBuffer {
             return nil
         }
 
-        let lenBytes: [UInt8] = self.readBytes(length: 3)!
```
Comment (weissi) [inline Sources/NIOHTTP2/HTTP2FrameParser.swift]: I think you should keep the old code but replace

```
let lenBytes: [UInt8] = self.readBytes(length: 3)!
```

with

```
let lenHigh = self.readInteger(as: UInt16.self)
let lenLow = self.readInteger(as: UInt8.self)
let len = UInt32(lenHigh) << 16 | UInt32(lenLow)
```

File: Sources/NIOHTTP2/HTTP2FrameParser.swift [hunk 2]
```diff
@@ -1163,13 +1163,14 @@ fileprivate extension ByteBuffer {
             return nil
```
Comment (weissi) [inline Sources/NIOHTTP2/HTTP2FrameParser.swift]: I don't really like the pre-calculation up here ^^^ and the `!`s down there. I think this should be replaced with

```swift
let saveSelf = self
guard let lenHigh = self.readInteger(as: UInt16.self),
      let lenLow = self.readInteger(as: UInt8.self),
      let lenLow = self.readInteger(as: UInt8.self),
      let type = self.readInteger(as: UInt8.self),
      let flags = self.readInteger(as: UInt8.self),
      let streamID = self.readInteger(as: UInt32.self) else {
    self = saveSelf
    return nil
}

return FrameHeader(...)
```

Comment (Lukasa) [reply Sources/NIOHTTP2/HTTP2FrameParser.swift]: I disagree.

Specifically, the form you proposed does repeated and unnecessary bounds checking, incurring a bunch of potentially unnecessary pointer dereferences. Done maliciously it substantially increases the work needed to parse a frame header by forcing a quadratic behaviour when parsing frame headers.

Comment (weissi) [reply Sources/NIOHTTP2/HTTP2FrameParser.swift]: There are no pointer dereferences.

Comment (Lukasa) [reply Sources/NIOHTTP2/HTTP2FrameParser.swift]: Sure there are, reading data from the `ByteBuffer` is a pointer dereference.

Comment (weissi) [reply Sources/NIOHTTP2/HTTP2FrameParser.swift]: the bounds checks are integer comparisons. I agree that we should add a way that can hoist the bounds checks but we don't have one today. Going into unsafe-land is not a good excuse to reduce a constant factor

Comment (weissi) [reply Sources/NIOHTTP2/HTTP2FrameParser.swift]: For the future, I want something like

```
self.read( integer(UInt16.self) <*> integer(UInt8.self) <*> integer(UInt8.self) <*> integer(UInt8.self) )
```

that does the bounds check only once. But we're not there yet.

File: Sources/NIOHTTP2/HTTP2FrameParser.swift [hunk 3]
```diff
@@ -1159,17 +1159,17 @@ fileprivate struct FrameHeader {
...
         }
 
-        let lenBytes: [UInt8] = self.readBytes(length: 3)!
-        let len = Int(lenBytes[0]) << 16 | Int(lenBytes[1]) << 8 | Int(lenBytes[2])
-        let type: UInt8 = self.readInteger()!
-        let flags: UInt8 = self.readInteger()!
-        let rawStreamID: UInt32 = self.readInteger()!
-
-        return FrameHeader(length: len, type: type, flags: FrameFlags(rawValue: flags), rawStreamID: rawStreamID)
+        return FrameHeader(length: Int(lenHigh) << 8 | Int(lenLow), type: type, flags: FrameFlags(rawValue: flags), rawStreamID: rawStreamID)
```
Comment (weissi) [inline Sources/NIOHTTP2/HTTP2FrameParser.swift]: aren't we overwriting 8 bits here?

Comment (Lukasa) [reply Sources/NIOHTTP2/HTTP2FrameParser.swift]: No: the UInt16 is converted to an `Int` and shifted left 8 bits. Then the UInt8 is converted to an Int and or'ed with it, fitting into the empty space from the previous left shift.

Comment (weissi) [reply Sources/NIOHTTP2/HTTP2FrameParser.swift]: @Lukasa oops, thanks, brain fart :D

**Pull request merged**